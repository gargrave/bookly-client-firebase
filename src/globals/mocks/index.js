import { authorMocks } from './data/authorMocks';
import { bookBucketMocks } from './data/bookBucketMocks';
import { bookMocks } from './data/bookMocks';
import { profileMocks } from './data/profileMocks';
import { userMocks } from './data/userMocks';

export {
  authorMocks,
  bookBucketMocks,
  bookMocks,
  profileMocks,
  userMocks,
};
