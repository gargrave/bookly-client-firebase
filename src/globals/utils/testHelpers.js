import React from 'react';
import { shallow } from 'enzyme';

export class ComponentBuilder {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = defaultProps;
  }

  shallowGetComponent(extraProps = {}) {
    const props = {
      ...this.defaultProps,
      ...extraProps,
    };
    return shallow(<this.component {...props} />);
  }
}
