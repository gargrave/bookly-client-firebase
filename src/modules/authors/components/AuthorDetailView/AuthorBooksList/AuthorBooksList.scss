@import '../../../../../styles/_index';

.authorBooksList {
  margin-top: 40px;
}

.header {
  @include centerAlignedHeader();
  color: $textColorLight;
  margin-bottom: 20px;
}
