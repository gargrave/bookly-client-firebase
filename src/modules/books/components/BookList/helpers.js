// @flow
import React, { Fragment } from 'react';

import type { Book, BookBucket } from '../../flowtypes';

import { filterAndBucket, filterBooksByTitle } from '../helpers';

import BookListBucket from '../BookListBucket/BookListBucket';
import BookListDetail from '../BookListDetail/BookListDetail';

export const bucketedBookList = (
  books: Book[],
  onBookClick: Function,
  filterBy?: string,
) => {
  return (
    <Fragment>
      {
        filterAndBucket(books, filterBy)
          .map((bucket: BookBucket) => {
            return (
              <BookListBucket
                bucket={bucket}
                key={bucket.author}
                onBookClick={onBookClick}
              />
            );
          })
      }
    </Fragment>
  );
};

export const flatBookList = (
  books: Book[],
  onBookClick: Function,
  filterBy?: string,
) => {
  return (
    <Fragment>
      {
        filterBooksByTitle(books, filterBy)
          .map((book: Book) => {
            return (
              <BookListDetail
                book={book}
                key={book.id}
                onClick={onBookClick.bind(null, book.id)}
                showAuthor={true}
              />
            );
          })
      }
    </Fragment>
  );
};
