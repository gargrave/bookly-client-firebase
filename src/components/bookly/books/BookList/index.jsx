// @flow
import React, { Fragment } from 'react';
import { array, bool, func, string } from 'prop-types';

import type { Book, BookBucket } from '../../../../modules/books/flowtypes';

import { filterAndBucket, filterBooksByTitle } from '../helpers';

import Alert from '../../../../modules/common/components/Alert/Alert';
import BookListBucket from '../BookListBucket/';
import BookListDetail from '../BookListDetail/BookListDetail';

type Props = {
  books: Book[],
  filterBy?: string,
  onBookClick: Function,
  groupBooksByAuthor?: boolean,
};

const bucketedBookList = (
  books: Book[],
  onBookClick: Function,
  filterBy?: string,
) => {
  return (
    <Fragment>
      {
        filterAndBucket(books, filterBy)
          .map((bucket: BookBucket) => {
            return (
              <BookListBucket
                bucket={bucket}
                key={bucket.author}
                onBookClick={onBookClick}
              />
            );
          })
      }
    </Fragment>
  );
};

const flatBookList = (
  books: Book[],
  onBookClick: Function,
  filterBy?: string,
) => {
  return (
    <Fragment>
      {
        filterBooksByTitle(books, filterBy)
          .map((book: Book) => {
            return (
              <BookListDetail
                book={book}
                key={book.id}
                onClick={onBookClick.bind(null, book.id)}
                showAuthor={true}
              />
            );
          })
      }
    </Fragment>
  );
};

const renderBookList = (
  books: Book[],
  onBookClick: Function,
  filterBy?: string,
  groupBooksByAuthor?: boolean,
) => {
  return groupBooksByAuthor
    ? bucketedBookList(books, onBookClick, filterBy)
    : flatBookList(books, onBookClick, filterBy);
};

const noBooksMessage = () => {
  return (
    <Alert
      message={'No Books created yet!'}
      type={'info'}
    />
  );
};

const BookList = ({
  books,
  filterBy,
  onBookClick,
  groupBooksByAuthor,
}: Props) => {
  return (
    books.length
      ? renderBookList(books, onBookClick, filterBy, groupBooksByAuthor)
      : noBooksMessage()
  );
};

BookList.propTypes = {
  books: array.isRequired,
  filterBy: string,
  onBookClick: func.isRequired,
  groupBooksByAuthor: bool,
};

export default BookList;
