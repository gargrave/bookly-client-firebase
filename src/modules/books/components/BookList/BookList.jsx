// @flow
import React from 'react';
import { array, bool, func, string } from 'prop-types';

import type { Book } from '../../flowtypes';

import { bucketedBookList, flatBookList } from './helpers';

import Alert from '../../../common/components/Alert/Alert';

type Props = {
  books: Book[],
  filterBy?: string,
  onBookClick: Function,
  groupBooksByAuthor?: boolean,
};

const renderBookList = (
  books: Book[],
  onBookClick: Function,
  filterBy?: string,
  groupBooksByAuthor?: boolean,
) => {
  return groupBooksByAuthor
    ? bucketedBookList(books, onBookClick, filterBy)
    : flatBookList(books, onBookClick, filterBy);
};

const noBooksMessage = () => {
  return (
    <Alert
      message={'No Books created yet!'}
      type={'info'}
    />
  );
};

const BookList = ({
  books,
  filterBy,
  onBookClick,
  groupBooksByAuthor,
}: Props) => (
  books.length
    ? renderBookList(books, onBookClick, filterBy, groupBooksByAuthor)
    : noBooksMessage()
);

BookList.propTypes = {
  books: array.isRequired,
  filterBy: string,
  onBookClick: func.isRequired,
  groupBooksByAuthor: bool,
};

export default BookList;
