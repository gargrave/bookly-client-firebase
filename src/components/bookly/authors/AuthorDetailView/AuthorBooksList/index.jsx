// @flow
import React from 'react';
import { array, func, object } from 'prop-types';

import type { Author, Book } from '../../../../../constants/flowtypes';

import { buildClasses } from '../../../../../utils/cssHelpers';

import BookList from '../../../books/BookList';
import Button from '../../../../common/Button';

import './styles.css';

type Props = {
  author: Author,
  books: Book[],
  onBookAddClick: Function,
  onBookClick: Function,
};

const prefixClass = (cls = '') => `author-books-list${cls}`;

const AuthorBooksList = ({
  author,
  books,
  onBookAddClick,
  onBookClick,
}: Props) => {
  const authorName = `${author.firstName} ${author.lastName}`;

  return (
    <div className={buildClasses(prefixClass())}>
      <h5 className={buildClasses(prefixClass('__header'))}>
        Books by {authorName}
        <Button
          onClick={onBookAddClick.bind(this, author)}
          text="Add"
          type="success"
        />
      </h5>
      <BookList
        books={books}
        onBookClick={onBookClick}
      />
    </div>
  );
};

AuthorBooksList.propTypes = {
  author: object.isRequired,
  books: array.isRequired,
  onBookAddClick: func.isRequired,
  onBookClick: func.isRequired,
};

export default AuthorBooksList;
