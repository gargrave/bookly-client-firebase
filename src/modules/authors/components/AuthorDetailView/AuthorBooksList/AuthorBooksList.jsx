// @flow
import React from 'react';
import { array, func, object } from 'prop-types';

import type { Book } from '../../../../books/flowtypes';
import type { Author } from '../../../flowtypes';

import { buildClasses } from '../../../../../utils/cssHelpers';

import BookList from '../../../../books/components/BookList/BookList';
import Button from '../../../../common/components/Button/Button';

import './AuthorBooksList.css';

type Props = {
  author: Author,
  books: Book[],
  onBookAddClick: Function,
  onBookClick: Function,
};

const prefixClass = (cls = '') => `author-books-list${cls}`;

const AuthorBooksList = ({
  author,
  books,
  onBookAddClick,
  onBookClick,
}: Props) => {
  const authorName = `${author.firstName} ${author.lastName}`;

  return (
    <div className={buildClasses(prefixClass())}>
      <h5 className={buildClasses(prefixClass('__header'))}>
        Books by {authorName}
        <Button
          onClick={onBookAddClick.bind(null, author)}
          text="Add"
          type="success"
        />
      </h5>
      <BookList
        books={books}
        onBookClick={onBookClick}
      />
    </div>
  );
};

AuthorBooksList.propTypes = {
  author: object.isRequired,
  books: array.isRequired,
  onBookAddClick: func.isRequired,
  onBookClick: func.isRequired,
};

export default AuthorBooksList;
