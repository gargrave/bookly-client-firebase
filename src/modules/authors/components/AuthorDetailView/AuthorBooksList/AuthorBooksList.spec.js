import { ComponentBuilder } from '../../../../../utils/testHelpers'

import { authorMocks, bookMocks } from '../../../../../globals/mocks/'

import BookList from '../../../../books/components/BookList/BookList'

import AuthorBooksList from './AuthorBooksList'

const defaultProps = {
  author: authorMocks[0],
  books: bookMocks.slice(-2),
  onBookAddClick: jest.fn(),
  onBookClick: jest.fn(),
}

const builder = new ComponentBuilder(AuthorBooksList, defaultProps)

describe.only('AuthorBooksList', () => {
  let component

  it('matches the snapshot', () => {
    component = builder.shallowGetComponent()
    expect(component).toMatchSnapshot()
  })

  it('renders correctly', () => {
    component = builder.shallowGetComponent()
    expect(component.find(BookList)).toHaveLength(1)
  })
})
