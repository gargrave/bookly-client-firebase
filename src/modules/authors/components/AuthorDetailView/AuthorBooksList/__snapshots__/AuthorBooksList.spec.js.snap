// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorBooksList matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthorBooksList
    author={
      Object {
        "bookCount": 1,
        "created": 2017-07-01T01:16:43.881Z,
        "firstName": "Aldous",
        "id": "oiw223h3qndlksadffl3ksdf",
        "lastName": "Huxley",
        "updated": 2017-07-01T01:16:43.881Z,
      }
    }
    books={
      Array [
        Object {
          "author": Object {
            "bookCount": 1,
            "created": 2017-07-06T22:02:11.504Z,
            "firstName": "Kurt",
            "id": "9328fhs938hfdsdblkhf98",
            "lastName": "Vonnegut",
            "updated": 2017-07-06T22:02:11.504Z,
          },
          "created": 2017-11-01T01:36:26.220Z,
          "id": "e2TedzRQJKJ1VL5m7QhG",
          "title": "Mother Night",
          "updated": 2017-11-01T01:36:26.220Z,
        },
        Object {
          "author": Object {
            "bookCount": 1,
            "created": 2017-07-06T22:02:11.504Z,
            "firstName": "Kurt",
            "id": "9328fhs938hfdsdblkhf98",
            "lastName": "Vonnegut",
            "updated": 2017-07-06T22:02:11.504Z,
          },
          "created": 2017-11-01T01:39:26.220Z,
          "id": "8ZYttrQY042IhlZ11BVJ",
          "title": "Breakfast of Champions",
          "updated": 2017-11-01T01:39:26.220Z,
        },
      ]
    }
    onBookAddClick={[Function]}
    onBookClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h5
          className={undefined}
        >
          Books by 
          Aldous Huxley
          <Button
            onClick={[Function]}
            text="Add"
            type="success"
          />
        </h5>,
        <BookList
          books={
            Array [
              Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:11.504Z,
                  "firstName": "Kurt",
                  "id": "9328fhs938hfdsdblkhf98",
                  "lastName": "Vonnegut",
                  "updated": 2017-07-06T22:02:11.504Z,
                },
                "created": 2017-11-01T01:36:26.220Z,
                "id": "e2TedzRQJKJ1VL5m7QhG",
                "title": "Mother Night",
                "updated": 2017-11-01T01:36:26.220Z,
              },
              Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:11.504Z,
                  "firstName": "Kurt",
                  "id": "9328fhs938hfdsdblkhf98",
                  "lastName": "Vonnegut",
                  "updated": 2017-07-06T22:02:11.504Z,
                },
                "created": 2017-11-01T01:39:26.220Z,
                "id": "8ZYttrQY042IhlZ11BVJ",
                "title": "Breakfast of Champions",
                "updated": 2017-11-01T01:39:26.220Z,
              },
            ]
          }
          onBookClick={[Function]}
        />,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Books by ",
            "Aldous Huxley",
            <Button
              onClick={[Function]}
              text="Add"
              type="success"
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          "Books by ",
          "Aldous Huxley",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "text": "Add",
              "type": "success",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "h5",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "books": Array [
            Object {
              "author": Object {
                "bookCount": 1,
                "created": 2017-07-06T22:02:11.504Z,
                "firstName": "Kurt",
                "id": "9328fhs938hfdsdblkhf98",
                "lastName": "Vonnegut",
                "updated": 2017-07-06T22:02:11.504Z,
              },
              "created": 2017-11-01T01:36:26.220Z,
              "id": "e2TedzRQJKJ1VL5m7QhG",
              "title": "Mother Night",
              "updated": 2017-11-01T01:36:26.220Z,
            },
            Object {
              "author": Object {
                "bookCount": 1,
                "created": 2017-07-06T22:02:11.504Z,
                "firstName": "Kurt",
                "id": "9328fhs938hfdsdblkhf98",
                "lastName": "Vonnegut",
                "updated": 2017-07-06T22:02:11.504Z,
              },
              "created": 2017-11-01T01:39:26.220Z,
              "id": "8ZYttrQY042IhlZ11BVJ",
              "title": "Breakfast of Champions",
              "updated": 2017-11-01T01:39:26.220Z,
            },
          ],
          "onBookClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h5
            className={undefined}
          >
            Books by 
            Aldous Huxley
            <Button
              onClick={[Function]}
              text="Add"
              type="success"
            />
          </h5>,
          <BookList
            books={
              Array [
                Object {
                  "author": Object {
                    "bookCount": 1,
                    "created": 2017-07-06T22:02:11.504Z,
                    "firstName": "Kurt",
                    "id": "9328fhs938hfdsdblkhf98",
                    "lastName": "Vonnegut",
                    "updated": 2017-07-06T22:02:11.504Z,
                  },
                  "created": 2017-11-01T01:36:26.220Z,
                  "id": "e2TedzRQJKJ1VL5m7QhG",
                  "title": "Mother Night",
                  "updated": 2017-11-01T01:36:26.220Z,
                },
                Object {
                  "author": Object {
                    "bookCount": 1,
                    "created": 2017-07-06T22:02:11.504Z,
                    "firstName": "Kurt",
                    "id": "9328fhs938hfdsdblkhf98",
                    "lastName": "Vonnegut",
                    "updated": 2017-07-06T22:02:11.504Z,
                  },
                  "created": 2017-11-01T01:39:26.220Z,
                  "id": "8ZYttrQY042IhlZ11BVJ",
                  "title": "Breakfast of Champions",
                  "updated": 2017-11-01T01:39:26.220Z,
                },
              ]
            }
            onBookClick={[Function]}
          />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Books by ",
              "Aldous Huxley",
              <Button
                onClick={[Function]}
                text="Add"
                type="success"
              />,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [
            "Books by ",
            "Aldous Huxley",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClick": [Function],
                "text": "Add",
                "type": "success",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "h5",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "books": Array [
              Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:11.504Z,
                  "firstName": "Kurt",
                  "id": "9328fhs938hfdsdblkhf98",
                  "lastName": "Vonnegut",
                  "updated": 2017-07-06T22:02:11.504Z,
                },
                "created": 2017-11-01T01:36:26.220Z,
                "id": "e2TedzRQJKJ1VL5m7QhG",
                "title": "Mother Night",
                "updated": 2017-11-01T01:36:26.220Z,
              },
              Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:11.504Z,
                  "firstName": "Kurt",
                  "id": "9328fhs938hfdsdblkhf98",
                  "lastName": "Vonnegut",
                  "updated": 2017-07-06T22:02:11.504Z,
                },
                "created": 2017-11-01T01:39:26.220Z,
                "id": "8ZYttrQY042IhlZ11BVJ",
                "title": "Breakfast of Champions",
                "updated": 2017-11-01T01:39:26.220Z,
              },
            ],
            "onBookClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
